fxrecorder/src/lib/recorder.rs

```diff
@@ -0,0 +1,172 @@
+    /// before the recording finishes.
+    ///
+    /// The path to the recording is returned.
+    async fn finish_recording(&self, handle: Self::Handle) -> Result<PathBuf, Self::Error>;
+}
+
+/// A Recorder that uses `ffmpeg`.
+pub struct FfmpegRecorder<'a> {
+    log: slog::Logger,
+    config: &'a RecordingConfig,
+}
+
+/// A handle for the [`FfmpegRecorder`][FfmpegRecorder]
+///
+/// [FfmpegRecorder]: struct.FfmpegRecorder.html
+pub struct FfmpegRecordingHandle {
+    ffmpeg: Child,
+    output_path: PathBuf,
+}
+
+impl<'a> FfmpegRecorder<'a> {
+    pub fn new(log: slog::Logger, config: &'a RecordingConfig) -> Self {
+        FfmpegRecorder { log, config }
+    }
+}
+
+/// An error recording with ffmpeg.
+#[derive(Debug, Error)]
+pub enum FfmpegRecordingError {
+    #[error("Could not start ffmpeg: {}", .0)]
+    Start(#[source] io::Error),
+
+    #[error("Could not ask ffmpeg to quit: {}", .0)]
+    WriteQ(#[source] io::Error),
+
+    #[error("could not wait for ffmpeg to exist: {}", .0)]
+    Wait(#[source] io::Error),
+
+    #[error("ffmpeg exited with nonzero status: {}", .0)]
+    ExitStatus(i32),
+}
+
+#[async_trait]
+impl<'a> Recorder for FfmpegRecorder<'a> {
+    type Handle = FfmpegRecordingHandle;
+    type Error = FfmpegRecordingError;
+
+    async fn start_recording(&self, recording_dir: &Path) -> Result<Self::Handle, Self::Error> {
+        let output_path = recording_dir.join("recording.mp4");
+        let input_arg = format!("video={}", self.config.device);
+        let video_size_arg = format!("{}x{}", self.config.video_size.x, self.config.video_size.y);
+        let framerate_arg = self.config.frame_rate.to_string();
+
+        let mut args: Vec<&OsStr> = vec![
+            OsStr::new("-f"),
+            OsStr::new("dshow"),
+            OsStr::new("-i"),
+            OsStr::new(&input_arg),
+            OsStr::new("-video_size"),
+            OsStr::new(&video_size_arg),
+            OsStr::new("-rtbufsize"),
+            OsStr::new(&self.config.buffer_size),
+            OsStr::new("-framerate"),
+            OsStr::new(&framerate_arg),
+        ];
+
+        let scale;
+        if let Some(ref output_size) = self.config.output_size {
+            scale = format!("scale={}:{}", output_size.x, output_size.y);
+
+            args.push(OsStr::new("-vf"));
+            args.push(OsStr::new(&scale));
+        }
+
+        args.push(output_path.as_os_str());
+
+        info!(self.log, "starting ffmpeg...");
+        let ffmpeg = Command::new(&self.config.ffmpeg_path)
+            .args(&args)
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped())
+            .spawn()
+            .map_err(FfmpegRecordingError::Start)?;
+
+        Ok(FfmpegRecordingHandle {
+            ffmpeg,
+            output_path,
+        })
+    }
+
+    async fn finish_recording(&self, mut handle: Self::Handle) -> Result<PathBuf, Self::Error> {
+        delay_for(Duration::from_secs(
```

Comment:
Nit: this feels like a slightly awkward place to put this. I would think that behavior like this would want to live in the code that calls `finish_recording`, especially given that it's not immediately obvious from the name `finish_recording` that it's going to wait for some number of seconds.

Reply:
I put this in here so that we specifically so we don't have to delay in the caller. If we delay in the caller, then we also are delaying in unit tests, which isn't a great test experience (and `cargo test` will yell at you).

Reply:
An alternative is that start_recording() can `spawn()` another task and `finish_recording` can simply wait for that task to complete. We can send a message on a channel to notify the task that it should start its timer, e.g. a trait that looks like:

```rust
#[async trait]
pub trait Recorder {
  type Handle;
  type Error: Error + 'static;

  // Spawn a new recording task.
  async fn start_recording(&self, directory: &Path) -> Result<Self::Handle, Self::Error>;

  // Notify the recording task that it should start its countdown timer (if any).
  // For testing this would be a no-op.
  async fn notify_firefox_started(&self, handle: &mut Self::Handle) -> Result<(), Self::Error>;

  // Attempt to join the underlying task handle.
  async fn finish_recording(&self, mut handle: Self::Handle) -> Result<PathBuf, Self::Error>;
}
```

and is used like: 
```rust
let mut handle = self
  .recorder
  .start_recording(tmpdir.path())
  .await
  .map_err(RecorderProtoError::Recording)?;

self.send(StartFirefox).await?;
self.recv::<StartedFirefox>().await?.unwrap();
self.recorder.notify_firefox_started(&mut handle).await.unwrap();
self.finish_recording(&self, handle).await.unwrap();
```

Reply:
My main issue is just that `finish_recording` does not give any indication of the delay. I would be content with simply renaming that to `finish_recording_after_delay` or similar?